### PR TITLE
Make LB options optional

### PR DESCRIPTION
**Pull request by bbbco**
I am getting this stack trace when I try to provision the `simple.rb` cluster from the example docs:

````
AWS::ELB::Errors::InvalidConfigurationRequest: load_balancer[webapp-elb] (@recipe_files::/home/bgoad/git/chef-provisioning-aws/docs/examples/simple.rb line 17) had an error: AWS::ELB::Errors::InvalidConfigurationRequest: Security group(s) can be applied to only an ELB in VPC.
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/aws-sdk-v1-1.59.1/lib/aws/core/client.rb:375:in `return_or_raise'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/aws-sdk-v1-1.59.1/lib/aws/core/client.rb:476:in `client_request'
(eval):3:in `create_load_balancer'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/aws-sdk-v1-1.59.1/lib/aws/elb/load_balancer_collection.rb:116:in `create'
/home/bgoad/git/chef-provisioning-aws/lib/chef/provisioning/aws_driver/driver.rb:77:in `block in allocate_load_balancer'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/mixin/why_run.rb:52:in `call'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/mixin/why_run.rb:52:in `add_action'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/provider.rb:156:in `converge_by'
/home/bgoad/.rvm/gems/ruby-2.1.0/bundler/gems/chef-provisioning-c9e700525163/lib/chef/provisioning/chef_provider_action_handler.rb:54:in `perform_action'
/home/bgoad/git/chef-provisioning-aws/lib/chef/provisioning/aws_driver/driver.rb:76:in `allocate_load_balancer'
/home/bgoad/.rvm/gems/ruby-2.1.0/bundler/gems/chef-provisioning-c9e700525163/lib/chef/provider/load_balancer.rb:30:in `block in <class:LoadBalancer>'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/provider/lwrp_base.rb:138:in `instance_eval'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/provider/lwrp_base.rb:138:in `block in action'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/provider.rb:121:in `run_action'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/resource.rb:648:in `run_action'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/runner.rb:49:in `run_action'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/runner.rb:81:in `block (2 levels) in converge'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/runner.rb:81:in `each'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/runner.rb:81:in `block in converge'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/resource_collection.rb:98:in `block in execute_each_resource'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/resource_collection.rb:96:in `execute_each_resource'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/runner.rb:80:in `converge'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/client.rb:345:in `converge'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/client.rb:431:in `do_run'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/client.rb:213:in `block in run'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/client.rb:207:in `fork'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/client.rb:207:in `run'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/application.rb:236:in `run_chef_client'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/application/client.rb:338:in `block in run_application'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/application/client.rb:327:in `loop'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/application/client.rb:327:in `run_application'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/lib/chef/application.rb:55:in `run'
/home/bgoad/.rvm/gems/ruby-2.1.0/gems/chef-11.16.4/bin/chef-client:26:in `<top (required)>'
/home/bgoad/.rvm/gems/ruby-2.1.0/bin/chef-client:23:in `load'
/home/bgoad/.rvm/gems/ruby-2.1.0/bin/chef-client:23:in `<main>'
/home/bgoad/.rvm/gems/ruby-2.1.0/bin/ruby_executable_hooks:15:in `eval'
/home/bgoad/.rvm/gems/ruby-2.1.0/bin/ruby_executable_hooks:15:in `<main>'
````

Investigating the code, the `allocate_load_balancer` in the chef-provisioning-aws driver passes on security_group, availability_zones, and listeners options by default. However, the aws-sdk code indicates these are all optional, and in fact security_groups is only available with a VPC. In the case above, I could not create a new ELB  because we don't use VPC. 

I updated the driver code to make passing these three options truly optional.

### CODE REVIEW

**Comment by bbbco** [comment]
FYI, I just signed the Opscode Contributer form

File: lib/chef/provisioning/aws_driver/driver.rb [hunk 1]
```diff
@@ -53,17 +52,22 @@ def self.canonicalize_url(driver_url, config)
...
+                       elsif !security_group_id.nil?
                          ec2.security_groups[security_group_id]
                        end
       availability_zones = lb_options[:availability_zones]
       listeners = lb_options[:listeners]
 
+      lb_optionals = {}
+      lb_optionals[:security_groups] = [security_group] if security_group
+      lb_optionals[:availability_zones] = availability_zones if availability_zones
+      lb_optionals[:listeners] = listeners
```
Comment (jkeiser) [inline lib/chef/provisioning/aws_driver/driver.rb]: I think you meant to make these optional too?

**Comment by jkeiser** [comment]
@bbbco one comment there--it looks like listeners are being treated differently from the other two.

**Comment by bbbco** [comment]
Thx @jkeiser for catching that. I fixed that change, and updated my fork, so the PR should be able to be merged in just fine.

File: lib/chef/provisioning/aws_driver/driver.rb [hunk 2]
```diff
@@ -58,17 +59,22 @@ def self.canonicalize_url(driver_url, config)
 
     # Load balancer methods
     def allocate_load_balancer(action_handler, lb_spec, lb_options, machine_specs)
-      security_group_name = lb_options[:security_group_name] || 'default'
+      security_group_name = lb_options[:security_group_name]
```
Comment (jkeiser) [inline lib/chef/provisioning/aws_driver/driver.rb]: Question: what happened to security_group_name defaulting to `'default'`?  Is that not needed anymore?

**Comment by bbbco** [comment]
Right. Including `security_groups` in the `elb.load_balancers.create()` function throws the `AWS::ELB::Errors::InvalidConfigurationRequest` because it should only be used with VPC. So my code sets each of the options to null and only passes them to the `elb.load_balancers.create()` function if they are explicitly set by the user.

**Comment by jkeiser** [comment]
Right.  We don't want a `nil` security group for sure.  But that's different from saying "the user must always specify a security group."  The default security group exists: http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-network-security.html#default-security-group

It seems reasonable that LBs would be put in the default security group--do you happen to know if that's the case?

Also, this is probably a simpler representation of what you wrote (though I'll get rid of the 'default' if we find out that it's, well, the default :)  I have this in my tree so if it works for you you I'm fine with committing it.

```ruby
      if lb_options[:security_group_id]
        security_group = ec2.security_groups[security_group_id]
      else
        security_group = ec2.security_groups.filter('group-name', lb_options[:security_group_name] || 'default')
      end
```

**Comment by jkeiser** [comment]
It looks like it is :)  In that case, here's the code:

```ruby
      if lb_options[:security_group_id]
        security_group = ec2.security_groups[:security_group_id]
      elsif lb_options[:security_group_name]
        security_group = ec2.security_groups.filter('group-name', lb_options[:security_group_name])
      end
```

**Comment by jkeiser** [comment]
(The reason I'm suggesting the different code is I found my brain twisting a little at the negations of negations--I *think* this will have the same effect.)

**Comment by jkeiser** [comment]
I wasn't sure I could do this, but I guess I can :)  The last commit here has my changes on top of your changes.  What a tangled web we weave!

https://github.com/bbbco/chef-provisioning-aws/pull/1

**Comment by bbbco** [comment]
Instances do have a `default` security group automatically applied to them upon creation. You can add other security groups as desired.

Load Balancers by default do not have the option to set security groups unless you use VPC.  Security groups created on EC2 cannot be applied to ELBs. If we try to use the aws-sdk to pass in a `default` security group we get an `AWS::ELB::Errors::InvalidConfigurationRequest`, which is the whole point of this PR.

If a user has their own VPCs, then they should be able to apply a security group from their VPC to the load balancer, since that option is available via the aws-sdk.

So the `security_groups` variable can be nil if the config is not set. Thus the security_groups config will only be passed in if explicitly configured on the load balancer config by the user.

Your code is definitely much simpler. I wasn't thinking about refactoring, just trying to make it work :)

**Comment by jkeiser** [comment]
I've merged this as 40a0d8ddac1c6d0c09a4a39f62b11b2400a598c4 .  @bbbco would you mind testing your stuff against master to make sure I haven't rebroken you?

**Comment by jkeiser** [comment]
Also, I appreciate the education, thanks tons :)

**Pull request closed without merging**